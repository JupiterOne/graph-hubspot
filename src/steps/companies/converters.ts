import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { Company } from '../../types';
import { getEntityKey } from '../../utils';
import { Entities } from '../constants';

function validURL(str) {
  const pattern = new RegExp(
    '^(https?:\\/\\/)' + // protocol
      '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|' + // domain name
      '((\\d{1,3}\\.){3}\\d{1,3}))' + // OR ip (v4) address
      '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' + // port and path
      '(\\?[;&a-z\\d%_.~+=-]*)?' + // query string
      '(\\#[-a-z\\d_]*)?$',
    'i',
  ); // fragment locator
  return !!pattern.test(str);
}

export function createCompanyEntity(data: Company) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.COMPANY._class,
        _type: Entities.COMPANY._type,
        _key: getEntityKey(Entities.COMPANY, data.companyId.toString()),
        id: data.companyId.toString(),
        portalId: data.portalId,
        name: data.properties.name?.value,
        archived: data.isDeleted,
        host: data.properties.website?.value,
        domain: data.properties.domain?.value,
        city: data.properties.city?.value,
        industry: data.properties.industry?.value,
        ownerId: data.properties.hubspot_owner_id?.value,
        public: data.properties.is_public?.value === 'true',
        website: validURL(data.properties.website?.value)
          ? data.properties.website?.value
          : undefined,
        external: !!data.properties.website?.value,
        createdOn: data.properties.createdate?.value
          ? parseInt(data.properties.createdate?.value, 10)
          : undefined,
        updatedOn: data.properties.hs_lastmodifieddate?.value
          ? parseInt(data.properties.hs_lastmodifieddate?.value, 10)
          : undefined,
      },
    },
  });
}
