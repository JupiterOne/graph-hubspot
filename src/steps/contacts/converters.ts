import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Contact } from '../../types';
import { getEntityKey } from '../../utils';
import { Entities } from '../constants';

export function createContactEntity(data: Contact) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.CONTACT._class,
        _type: Entities.CONTACT._type,
        _key: getEntityKey(Entities.CONTACT, data.id),
        archived: data.archived,
        createdOn: parseTimePropertyValue(data.createdAt),
        updatedOn: parseTimePropertyValue(data.updatedAt),
        email: data.properties.email,
        name: `${data.properties.firstname} ${data.properties.lastname}`,
        username: data.properties.email,
      },
    },
  });
}
