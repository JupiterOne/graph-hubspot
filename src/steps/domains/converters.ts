import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Domain } from '../../types';
import { getEntityKey } from '../../utils';
import { Entities } from '../constants';

export function createDomainEntity(data: Domain) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.DOMAIN._class,
        _type: Entities.DOMAIN._type,
        _key: getEntityKey(Entities.DOMAIN, data.id),
        domainName: data.domain,
        name: data.domain,
        createdOn: parseTimePropertyValue(data.createdAt),
        updatedOn: parseTimePropertyValue(data.updatedAt),
      },
    },
  });
}
