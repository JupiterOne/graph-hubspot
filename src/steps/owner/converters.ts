import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { Owner } from '../../types';

export function createOwnerEntity(owner: Owner) {
  return createIntegrationEntity({
    entityData: {
      source: {
        id: owner.id,
      },
      assign: {
        _key: `hubspot_owner:${owner.id}`,
        _type: 'acme_user',
        _class: 'User',
        // This is a custom property that is not a part of the data model class
        // hierarchy. See: https://github.com/JupiterOne/data-model/blob/master/src/schemas/User.json
        firstName: owner.firstName,
        // lastName: owner.lastName,
        name: `${owner.firstName} ${owner.lastName}`,
        email: owner.email,
        username: owner.email,
      },
    },
  });
}
