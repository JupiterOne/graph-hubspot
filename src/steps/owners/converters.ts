import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Owner } from '../../types';
import { getEntityKey } from '../../utils';
import { Entities } from '../constants';

export function createUserEntity(data: Owner) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.USER._class,
        _type: Entities.USER._type,
        _key: getEntityKey(Entities.USER, data.id),
        id: data.id,
        name:
          data.firstName && data.lastName
            ? `${data.firstName} ${data.lastName}`
            : data.email,
        firstName: data.firstName,
        lastName: data.lastName,
        email: data.email,
        username: data.email,
        archived: data.archived,
        userId: data.userId,
        createdOn: parseTimePropertyValue(data.createdAt),
        updatedOn: parseTimePropertyValue(data.updatedAt),
      },
    },
  });
}
