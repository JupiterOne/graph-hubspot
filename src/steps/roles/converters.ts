import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { Role } from '../../types';
import { getEntityKey } from '../../utils';
import { Entities } from '../constants';

export function createRoleEntity(data: Role) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.ROLE._class,
        _type: Entities.ROLE._type,
        _key: getEntityKey(Entities.ROLE, data.id),
        name: data.name,
        requiresBillingWrite: data.requiresBillingWrite,
      },
    },
  });
}
