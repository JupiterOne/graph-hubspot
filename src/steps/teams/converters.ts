import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  Relationship,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import { Team } from '../../types';
import { getEntityKey } from '../../utils';
import { Entities, Relationships } from '../constants';

export function createTeamEntity(data: Team) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.TEAM._class,
        _type: Entities.TEAM._type,
        _key: getEntityKey(Entities.TEAM, data.id),
        name: data.name,
      },
    },
  });
}

export function createOwnerTeamRelationship(
  owner: Entity,
  team: Entity,
): Relationship {
  const parentKey = owner._key;
  const childKey = team._key;

  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    fromKey: parentKey,
    fromType: Entities.USER._type,
    toKey: childKey,
    toType: Entities.TEAM._type,
    properties: {
      _type: Relationships.USER_HAS_TEAM._type,
    },
  });
}
