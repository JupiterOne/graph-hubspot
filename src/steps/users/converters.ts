import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { User } from '../../types';
import { getEntityKey } from '../../utils';
import { Entities } from '../constants';

export function createUserEntity(data: User) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.USER._class,
        _type: Entities.USER._type,
        _key: getEntityKey(Entities.USER, data.id),
        id: data.id,
        username: data.email,
        name: data.email,
        email: data.email,
        roleId: data.roleId,
        primaryTeamId: data.primaryTeamId,
      },
    },
  });
}
