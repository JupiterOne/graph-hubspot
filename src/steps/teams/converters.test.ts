import { getMockTeam } from '../../../test/mocks';
import { createTeamEntity } from './converters';

describe('#createTeamEntity', () => {
  test('should convert to entity', () => {
    expect(createTeamEntity(getMockTeam())).toMatchSnapshot();
  });
});
