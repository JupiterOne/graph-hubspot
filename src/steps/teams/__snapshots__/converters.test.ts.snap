// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createTeamEntity should convert to entity 1`] = `
Object {
  "_class": Array [
    "Team",
  ],
  "_key": "hubspot_team:81106262",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "id": "81106262",
        "name": "Team A",
      },
    },
  ],
  "_type": "hubspot_team",
  "createdOn": undefined,
  "displayName": "Team A",
  "id": "81106262",
  "name": "Team A",
}
`;
